### ISSUE TITLE
Watch flushing in chunked encoder may be inefficient at scale

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#24729]
We currently invoke flush on the HTTP response after every watch event, in order to guarantee clients see watches immediately.  That flush impacts throughput at scale, and may result in undesirable context switches in the server process.  We should investigate deferred flushes, allowing the client to set an upper latency bound, or automatically adjusting the flush rate due to throughput.

In a 8 core benchmark of watch decode performance, it's roughly 25% faster not to flush than to flush (30us per iteration vs 40us per iteration) - and time spent over 1m in that loop goes from 67% waiting on syscall to 17% waiting on sys call, with much more time spent in GC.

@wojtek-t @kubernetes/sig-scalability

**Comment by hongchaodeng** [comment]
@smarterclayton 
By "deferred flush" do you mean batching?
My understanding is that events are changes happened on objects. Is status changes (e.g. Pod pending -> running) counted?

**Comment by wojtek-t** [comment]
>
My understanding is that events are changes happened on objects. Is status changes (e.g. Pod pending -> running) counted?

Yes - all changes on all objects.

>
By "deferred flush" do you mean batching?

It means that if you have a lot of changes it's probably good enough to avoid e.g. 5ms and flush all that were created during those 5ms than to flush them after each event. (Note that this 5ms is some random number, we probably need more thinking how often exactly we should do it).

I agree that some load-related think would be the best, but I think we won't get to it in 1.3.

**Comment by smarterclayton** [comment]
Within the watch code it's fairly easy to do counts and load based tracking
(it's a for loop with a select).  We *could* pretty easily track the
instantaneous rate of the watch and alter the flush interval pretty
efficiently.  Will put on the "nice to have for 1.3" list if someone gets
some free time.

**Comment by smarterclayton** [comment]
My only worry is that timer creation would actually limit the benefit (an
extra timer per watch).

**Comment by smarterclayton** [comment]
Created a prototype / almost complete in #24768. For throughput bound watches (>20qps) this should offer a substantial benefit.
